### PR TITLE
Enable safetensors conversion from PyTorch to other frameworks without the torch requirement

### PR DESCRIPTION
**Pull request by LysandreJik**
This removes the need to have `torch` installed to proceed to a safetensors (serialized from PyTorch) checkpoint converted into Flax.

This therefore, enables loading any model in Flax that has a safetensors file on the Hub.

### CODE REVIEW

File: tests/test_modeling_flax_utils.py [hunk 1]
```diff
@@ -265,7 +264,6 @@ def test_safetensors_load_from_hub_from_safetensors_pt(self):
         safetensors_model = FlaxBertModel.from_pretrained("hf-internal-testing/tiny-bert-pt-safetensors")
         self.assertTrue(check_models_equal(flax_model, safetensors_model))
 
-    @require_torch
```
Comment (LysandreJik) [inline tests/test_modeling_flax_utils.py:268]: Removing these requirements is equivalent to testing the two codepaths that were tested by the two tests that are removed below

File: setup.py [hunk 2]
```diff
@@ -158,7 +158,7 @@
     "ruff>=0.1.5,<=0.2",
     "sacrebleu>=1.4.12,<2.0.0",
     "sacremoses",
-    "safetensors>=0.3.1",
+    "safetensors==0.4.1rc1",
```
Comment (LysandreJik) [inline setup.py:161]: This will need to change before merge

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 3]
```diff
@@ -148,21 +152,24 @@ def is_key_or_prefix_key_in_dict(key: Tuple[str]) -> bool:
...
+            bfloat16_dtype = "bfloat16"
+
+        state_dict = {}
+        for k, v in pt_state_dict.items():
+            if v.dtype == bfloat16_dtype:
+                state_dict[k] = v.float().numpy() if from_bin else v
+            else:
+                state_dict[k] = v.numpy() if from_bin else v
+
+        return state_dict
```
Comment (LysandreJik) [inline src/transformers/modeling_flax_pytorch_utils.py:170]: @Narsil, curious if you have something a bit cleaner than that in mind to do dtype analysis across safetensors + pt

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 4]
```diff
@@ -224,15 +230,10 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
                 continue
 
             # also add unexpected weight so that warning is thrown
-            flax_state_dict[("params",) + flax_key] = (
-                jnp.asarray(flax_tensor) if not is_bfloat_16 else jnp.asarray(flax_tensor, dtype=jnp.bfloat16)
-            )
-
+            flax_state_dict[("params",) + flax_key] = np.asarray(flax_tensor)
```
Comment (amyeroberts) [inline src/transformers/modeling_flax_pytorch_utils.py:233]: Why is this one set as a numpy array rather than jnp?

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

File: src/transformers/modeling_flax_pytorch_utils.py (hunk 4 above)
Comment (LysandreJik) [reply src/transformers/modeling_flax_pytorch_utils.py:233]: oops yes I think it should be better

File: setup.py (hunk 2 above)
Comment (Narsil) [reply setup.py:161]: Release ready.

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for working on this! Just a small suggestion from me regarding the implicit upcasting of `bfloat16` pt params, but otherwise looks good 👍

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 5]
```diff
@@ -224,15 +230,10 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
                 continue
 
             # also add unexpected weight so that warning is thrown
-            flax_state_dict[("params",) + flax_key] = (
-                jnp.asarray(flax_tensor) if not is_bfloat_16 else jnp.asarray(flax_tensor, dtype=jnp.bfloat16)
```
[the remark this answers is not in the document]
Comment (LysandreJik) [reply src/transformers/modeling_flax_pytorch_utils.py:228]: Thanks for highlighting it! It was handled correctly when loading from a safetensors but not when loading from a bin file. I've added tests to ensure it works as expected in 6135334

File: src/transformers/modeling_flax_pytorch_utils.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/modeling_flax_pytorch_utils.py]: Not really torch.bfloat16 is really singular compared to numpy dtypes (which is a custom dtype for flax/tf)

**Comment by LysandreJik** [comment]
Would still like a review @sanchit-gandhi if you have the time for it :ok_man:

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 6]
```diff
@@ -148,21 +152,28 @@ def is_key_or_prefix_key_in_dict(key: Tuple[str]) -> bool:
...
+    def numpify_state_dict(_pt_state_dict):
+        from_bin = is_torch_available() and isinstance(next(iter(_pt_state_dict.values())), torch.Tensor)
+        _state_dict = {}
+        for k, v in _pt_state_dict.items():
+            if is_bfloat16(v):
+                _state_dict[k] = v.float().numpy() if from_bin else v
+            else:
+                _state_dict[k] = v.numpy() if from_bin else v
+
+        return _state_dict
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_pytorch_utils.py:174]: Suggestion to remove the two shallow functions and perform everything in the main body of the function. To me this is clearer than the two new function definitions!
```suggestion
    from_bin = is_torch_available() and isinstance(next(iter(pt_state_dict.values())), torch.Tensor)
    bfloat16 = torch.bfloat16 if from_bin else "bfloat16"

    for k, v in pt_state_dict.items():
        if v.dtype == bfloat16:
            pt_state_dict[k] = v.float().numpy() if from_bin else v
        else:
            pt_state_dict[k] = v.numpy() if from_bin else v
```

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 7]
```diff
@@ -180,16 +191,16 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
...
-        model_prefix not in {k.split(".")[0] for k in pt_state_dict.keys()}
+        model_prefix not in {k.split(".")[0] for k in state_dict.keys()}
     )
 
     # Need to change some parameters name to match Flax names
-    for pt_key, pt_tensor in pt_state_dict.items():
+    for pt_key, pt_tensor in state_dict.items():
         pt_tuple_key = tuple(pt_key.split("."))
-        is_bfloat_16 = weight_dtypes[pt_key] == torch.bfloat16
+        is_bfloat_16 = is_bfloat16(pt_state_dict[pt_key])
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_pytorch_utils.py:203]: And then this drops out to:
```suggestion
        is_bfloat_16 = pt_state_dict[pt_key].dtype == bfloat16
```

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 8]
```diff
@@ -180,16 +191,16 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
     flax_state_dict = {}
 
     load_model_with_head_into_base_model = (model_prefix not in flax_model_params) and (
-        model_prefix in {k.split(".")[0] for k in pt_state_dict.keys()}
+        model_prefix in {k.split(".")[0] for k in state_dict.keys()}
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_pytorch_utils.py:194]:
```suggestion
        model_prefix in {k.split(".")[0] for k in pt_state_dict.keys()}
```

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 9]
```diff
@@ -180,16 +191,16 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
...
     )
     load_base_model_into_model_with_head = (model_prefix in flax_model_params) and (
-        model_prefix not in {k.split(".")[0] for k in pt_state_dict.keys()}
+        model_prefix not in {k.split(".")[0] for k in state_dict.keys()}
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_pytorch_utils.py:197]:
```suggestion
        model_prefix not in {k.split(".")[0] for k in pt_state_dict.keys()}
```

File: src/transformers/modeling_flax_pytorch_utils.py [hunk 10]
```diff
@@ -180,16 +191,16 @@ def convert_pytorch_state_dict_to_flax(pt_state_dict, flax_model):
...
     )
 
     # Need to change some parameters name to match Flax names
-    for pt_key, pt_tensor in pt_state_dict.items():
+    for pt_key, pt_tensor in state_dict.items():
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_pytorch_utils.py:201]:
```suggestion
    for pt_key, pt_tensor in pt_state_dict.items():
```

**Review by sanchit-gandhi — APPROVED** [review]
Very nice, thanks for finishing this update @LysandreJik! And apologies for the late review here. Just a minor formatting suggestion from me, but otherwise functionality looks spot on 👌

**Pull request merged**